src/backend/metalll/src/factory.rs

```diff
@@ -642,43 +659,83 @@ impl CoreFactory<Resources> for Factory {
         unsafe { n::dispatch_release(semaphore.0) }
     }
 
-    // Emulated heap implementations
-    #[cfg(not(feature = "native_heap"))]
     fn create_heap(&mut self, heap_type: &HeapType, size: u64) -> n::Heap {
-        n::Heap { heap_type: *heap_type, size }
+        let (storage, cache) = map_heap_properties_to_storage_and_cache(heap_type.properties);
+
+        // Heaps cannot be used for CPU coherent resources
+        if self.private_caps.resource_heaps && storage != MTLStorageMode::Shared {
+            let descriptor = MTLHeapDescriptor::new();
+            descriptor.set_storage_mode(storage);
+            descriptor.set_cpu_cache_mode(cache);
+            descriptor.set_size(size);
+            n::Heap::Native(self.device.new_heap(descriptor))
+        } else {
+            n::Heap::Emulated { heap_type: *heap_type, size }
+        }
     }
-    #[cfg(not(feature = "native_heap"))]
+
     fn destroy_heap(&mut self, heap: n::Heap) {
+        match heap {
+            n::Heap::Emulated { .. } => {},
+            n::Heap::Native(heap) => unsafe { heap.release(); },
+        }
     }
 
-    #[cfg(not(feature = "native_heap"))]
     fn create_buffer(&mut self, size: u64, _stride: u64, _usage: buffer::Usage) -> Result<n::UnboundBuffer, buffer::CreationError> {
-        // TODO: map usage
-        Ok(n::UnboundBuffer(self.device.new_buffer(size, MTLResourceOptions::empty())))
+        Ok(n::UnboundBuffer {
+            size
+        })
     }
 
-    #[cfg(not(feature = "native_heap"))]
     fn get_buffer_requirements(&mut self, buffer: &n::UnboundBuffer) -> memory::MemoryRequirements {
+        // We don't know what memory type the user will try to allocate the buffer with, so we test them
+        // all get the most stringent ones. Note we don't check Shared because heaps can't use it
+        let mut max_size = 0;
+        let mut max_alignment = 0;
+        for &options in [
+            MTLResourceStorageModeManaged,
+            MTLResourceStorageModeManaged | MTLResourceCPUCacheModeWriteCombined,
+            MTLResourceStorageModePrivate,
+        ].iter() {
+            let requirements = self.device.heap_buffer_size_and_align(buffer.size, options);
+            max_size = cmp::max(max_size, requirements.size);
+            max_alignment = cmp::max(max_alignment, requirements.align);
+        }
         memory::MemoryRequirements {
-            size: buffer.0.length(),
-            alignment: 1,
+            size: max_size,
+            alignment: max_alignment,
         }
     }
 
-    #[cfg(not(feature = "native_heap"))]
     fn bind_buffer_memory(&mut self, heap: &n::Heap, offset: u64, buffer: n::UnboundBuffer) -> Result<n::Buffer, buffer::CreationError> {
-        Ok(n::Buffer(buffer.0))
+        let bound_buffer;
+        match *heap {
```

Comment:
similarly, please convert to functional style aka `let bound_buffer = match *heap {...};`